### PR TITLE
Exit as late as possible

### PR DESCRIPTION
**Pull request by greg0ire**
People might want to register other shutdown functions that should be
able to control the exit code themselves, without the deprecation error
handler taking over. The php manual says:

> If you call exit() within one registered shutdown function, processing
> will stop completely and no other registered shutdown functions will be
> called.

See https://secure.php.net/manual/en/function.register-shutdown-function.php

| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

### CODE REVIEW

**Comment by greg0ire** [comment]
@alcaeus please review

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: what would be even better would be to defer the whole close, so that deprecations triggered at shutdown time are also caught (and add a related test of course :) )

Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: The downside being that if a shutdown function uses `exit` and the whole closure is deferred too, then you won't get any deprecations. But I think catching shutdown time deprecations might outweight this, what do you think? If we want to go crazy, we may even reset the state after the first, non deferred display and if any deprecations are caught while shutting down, then display them in the deferred closure.

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: as you, both are fine to me :)

Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: I'll see what I can manage

**Comment by greg0ire** [comment]
Best reviewed when [ignoring whitespace](https://github.com/symfony/symfony/pull/26012/files?w=1)

@nicolas-grekas the second commit feels more like a feature than a bugfix to me. Would you be ok with moving it to a separate PR?

File: src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/shutdown_deprecations.phpt
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/shutdown_deprecations.phpt]: Tell me if you want more emphasis here, and if I should also add some header for the previous section.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Ended up "going crazy", tell me what you think

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: This calls for a data structure refactoring

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: let's do a code-style fix for now :) the "if" should be on one line

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: if there are shutdown-time deprecations, shouldn't this fail also?

File: src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/shutdown_deprecations.phpt
Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/shutdown_deprecations.phpt]: not more than existing titles, so as is LGTM

**Comment by alcaeus** [comment]
> the second commit feels more like a feature than a bugfix to me

TBH, I'd still consider it a bugfix: the original bug was that any existing deprecations cause other shutdown handlers to be ignored. This is fixed in your first commit. Fixing that bug exposes another bug, namely that any deprecations happening in other shutdown handlers don't reported. This is fixed in your second commit.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: "fixed" :P

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: done

**Review by nicolas-grekas — APPROVED** [review]
(failures should be fixed once merged up to master, isn't it?)

**Comment by greg0ire** [comment]
Let me try to test it on 5.4 to be 100% sure.

**Comment by greg0ire** [comment]
Don't merge, I get this: 

`"Use of undefined constant ARRAY_FILTER_USE_KEY - assumed 'ARRAY_FILTER_USE_KEY'"`

**Comment by greg0ire** [comment]
I fixed it however I could

**Comment by greg0ire** [comment]
Why exactly do we have those `Count` indices, BTW? I suppose the answer is performance, somehow?

**Comment by greg0ire** [comment]
Tested with `docker run -it -v $PWD:/tmp/sf squallcx/docker-php54-apache /bin/sh`

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Small tweak: s/Deprecations during shutdown/Shutdown-time deprecations:/

**Comment by nicolas-grekas** [comment]
Thank you @greg0ire.

**Pull request merged**
